subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/JavaGradlePluginPlugin.java

```diff
@@ -193,18 +193,22 @@ private TaskProvider<PluginUnderTestMetadata> createAndConfigurePluginUnderTestM
 
             pluginUnderTestMetadataTask.getOutputDirectory().set(project.getLayout().getBuildDirectory().dir(pluginUnderTestMetadataTask.getName()));
 
-            pluginUnderTestMetadataTask.getPluginClasspath().from((Callable<FileCollection>) () -> {
-                SourceSet sourceSet = extension.getPluginSourceSet();
-                Configuration runtimeClasspath = project.getConfigurations().getByName(sourceSet.getRuntimeClasspathConfigurationName());
-                ArtifactView view = runtimeClasspath.getIncoming().artifactView(config -> {
-                    config.componentFilter(spec(JavaGradlePluginPlugin::excludeGradleApi));
-                });
-                return pluginUnderTestMetadataTask.getProject().getObjects().fileCollection().from(
-                    sourceSet.getOutput(),
-                    view.getFiles().getElements()
-                );
-            });
+            pluginUnderTestMetadataTask.getPluginClasspath().from(
+                classpathForPlugin(pluginUnderTestMetadataTask, project, extension)
```

Comment:
How come this has been made eager?

Reply:
Isn't the new version more or less equivalent to the old one?
The `sourceSet`'s output is not going to be reassigned and is lazy by itself.  The configuration which produces the lazy artifact view should also be safe to access by name eagerly, unless it is removed from the container afterwards. 
The goal here was to avoid the `getProject` call inside the lambda. We can make the new implementation a little lazier by not creating the artifact view eagerly, though.

Reply:
I think I may be mistaken. I was worried that `view.getFiles()` was called outside of the callable, but looking into it more, it seems that is a lazy API.

Reply:
UPD: I see, we access the extension's `getSourceSet()` eagerly now, and it can be reassigned.

Reply:
Fixed in https://github.com/gradle/gradle/pull/23210/commits/3a0511f684e4272ef4f1b3a75664bfe7e115305a

Reply:
Is it okay to use the project instance itself in the lambda? How is calling `getProject` different from accessing an existing project instance?

Reply:
Yes, I think it's OK regarding the configuration cache contract: the project instance is only used to evaluate the file collection contents and is not carried to the execution time state. The difference here, I believe, is only in the build that produce the configuration cache entry: this build by itself still runs from the original project model, and its tasks might evaluate their inputs at the execution time, at which accessing the task's `getProject` is illegal, while using the captured project instance is still fine.